webapp/src/ts/modules/contacts/contacts.component.ts

```diff
@@ -428,22 +451,166 @@ export class ContactsComponent implements OnInit, OnDestroy{
+          openSendMessageModal: (sendTo) => this.openSendMessageModal(sendTo)
+        });
+
+        this.subscribeToAllContactXmlForms();
+        this.subscribeToSelectedContactXmlForms();
+      });
+    this.subscription.add(subscription);
+  }
+
+  private openContactMutedModal(form) {
+    if (!form.showUnmuteModal) {
+      this.router.navigate(['/contacts', this.selectedContact._id, 'report', form.code]);
+      return;
+    }
+
+    this.modalService
+      .show(ContactsMutedComponent)
+      .then(() => this.router.navigate(['/contacts', this.selectedContact._id, 'report', form.code]))
+      .catch(() => {});
+  }
+
+  private openSendMessageModal(sendTo) {
+    this.modalService
+      .show(SendMessageComponent, { initialState: { fields: { to: sendTo } } })
+      .catch(() => {});
+  }
+
+  private async setChildTypesBySelectedContact() {
+    if (!this.selectedContact) {
+      this.childTypesBySelectedContact = [];
+      return;
+    }
+
+    if (!this.selectedContact.type) {
+      const type = this.selectedContact.doc?.contact_type || this.selectedContact.doc?.type;
+      console.error(`Unknown contact type "${type}" for contact "${this.selectedContact.doc?._id}"`);
+      this.childTypesBySelectedContact = [];
+      return;
+    }
+
+    this.childTypesBySelectedContact = await this.getChildren(this.selectedContact.type.id);
+  }
+
+  private subscribeToAllContactXmlForms() {
+    if (this.subscriptionAllContactForms) {
+      this.subscriptionAllContactForms.unsubscribe();
+    }
+
+    this.subscriptionAllContactForms = this.xmlFormsService.subscribe(
+      'ContactForms',
+      { contactForms: true },
+      (error, forms) => {
+        if (error) {
+          console.error('Error fetching allowed contact forms', error);
+          return;
+        }
+
+        this.allowedChildPlaces = this.filterAllowedChildType(forms, this.childPlaces);
+        this.setLeftActionBar();
+        const allowedChildTypesBySelectedContact = this.filterAllowedChildType(forms, this.childTypesBySelectedContact);
+        this.globalActions.updateRightActionBar({
+          childTypes: this.getModelsFromChildTypes(allowedChildTypesBySelectedContact)
+        });
+      }
+    );
+    this.subscription.add(this.subscriptionAllContactForms);
+  }
+
+  private subscribeToSelectedContactXmlForms() {
+    if (!this.selectedContact) {
+      return;
+    }
+
+    if (this.subscriptionSelectedContactForms) {
+      this.subscriptionSelectedContactForms.unsubscribe();
+    }
+
+    this.subscriptionSelectedContactForms = this.xmlFormsService.subscribe(
+      'ContactList',
+      {
+        doc: this.selectedContact.doc,
+        contactSummary: this.selectedContact.summary?.context,
+        contactForms: false,
+      },
+      (error, forms) => {
+        if (error) {
+          console.error('Error fetching relevant forms', error);
+          return;
+        }
+
+        const formSummaries = forms
+          ?.map(xForm => {
```

Comment:
This notation looks very strange. Can we just return early if there are no forms?

Reply:
agreed, done